platform/linux-generic/arch/arm/odp_atomic.h

```diff
@@ -28,4 +27,51 @@ do {							\
 
 #endif  /* CONFIG_DMBSTR */
 
+/**
+ * @internal
+ * 128 bit store operation expression for the ATOMIC_OP macro
+ */
+#define ATOMIC_STORE_OP_128(new_val) \
+({ \
+	(_atom)->v = (new_val); \
+})
+
+/**
+ * @internal
+ * 128 bit CAS operation expression for the ATOMIC_OP macro
+ */
+#define ATOMIC_CAS_OP_128(ret_ptr, old_val, new_val) \
+({ \
+	int *_ret_ptr = ret_ptr; \
+	odp_u128_t *_old_val = old_val; \
+	odp_u128_t _new_val = new_val; \
+	if (((_atom)->v.u64[0] == (_old_val)->u64[0]) && \
+	    ((_atom)->v.u64[1] == (_old_val)->u64[1])) { \
+		(_atom)->v = (_new_val); \
+		*(_ret_ptr) = 1; \
+	} else { \
+		*(_ret_ptr) = 0; \
+	} \
+})
+
+/**
+ * @internal
+ * Helper macro for lock-based atomic operations on 128-bit integers
+ * @param[in,out] atom Pointer to the 128-bit atomic variable
+ * @param expr Expression used update the variable.
+ * @return The old value of the variable.
+ */
+#define ATOMIC_OP_128(atom, expr) \
+({ \
+	odp_u128_t _old_val; \
+	odp_atomic_u128_t *_atom = atom; \
+	/* Loop while lock is already taken, stop when lock becomes clear */ \
+	while (__atomic_test_and_set(&(_atom)->lock, __ATOMIC_ACQUIRE)) \
+		(void)0; \
+	_old_val = (_atom)->v; \
+	(expr); /* Perform whatever update is desired */ \
+	__atomic_clear(&(_atom)->lock, __ATOMIC_RELEASE); \
+	_old_val; /* Return old value */ \
+})
+
```

Comment:
To reduce the amount of duplicate code these macros for lock-based operation could be moved to `platform/linux-generic/odp_atomic.c`.